client/pluginmanager/drivermanager/manager.go

```diff
@@ -0,0 +1,286 @@
+}
+
+// New returns a new driver manager
+func New(c *Config) *manager {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &manager{
+		logger:          c.Logger.Named("driver_mgr"),
+		state:           c.State,
+		ctx:             ctx,
+		cancel:          cancel,
+		loader:          c.Loader,
+		pluginConfig:    c.PluginConfig,
+		updater:         c.Updater,
+		instances:       make(map[string]*instanceManager),
+		reattachConfigs: make(map[loader.PluginID]*shared.ReattachConfig),
+		allowedDrivers:  c.AllowedDrivers,
+		blockedDrivers:  c.BlockedDrivers,
+	}
+}
+
+// PluginType returns the type of plugin this mananger mananges
+func (*manager) PluginType() string { return base.PluginTypeDriver }
+
+// Run starts the mananger, initializes driver plugins and blocks until Shutdown
+// is called.
+func (m *manager) Run() {
+	// Load any previous plugin reattach configuration
+	m.loadReattachConfigs()
+
+	// Get driver plugins
+	driversPlugins := m.loader.Catalog()[base.PluginTypeDriver]
+	if len(driversPlugins) == 0 {
+		m.logger.Debug("exiting since there are no driver plugins")
+		m.cancel()
+		return
+	}
+
+	var availDrivers []string
+	var skippedDrivers []string
+	for _, d := range driversPlugins {
+		id := loader.PluginInfoID(d)
+		// Skip drivers that are not in the allowed list if it is set.
+		if _, ok := m.allowedDrivers[id.Name]; len(m.allowedDrivers) > 0 && !ok {
+			skippedDrivers = append(skippedDrivers, id.Name)
+			continue
+		}
+		// Skip fingerprinting drivers that are in the blocked list
+		if _, ok := m.blockedDrivers[id.Name]; ok {
+			skippedDrivers = append(skippedDrivers, id.Name)
+			continue
+		}
+
+		storeFn := func(c *plugin.ReattachConfig) error {
+			id := id
+			return m.storePluginReattachConfig(id, c)
+		}
+		fetchFn := func() (*plugin.ReattachConfig, bool) {
+			id := id
+			return m.fetchPluginReattachConfig(id)
+		}
+
+		instance := newInstanceManager(&instanceManagerConfig{
+			Logger:               m.logger,
+			Ctx:                  m.ctx,
+			Loader:               m.loader,
+			StoreReattach:        storeFn,
+			FetchReattach:        fetchFn,
+			PluginConfig:         m.pluginConfig,
+			ID:                   &id,
+			UpdateNodeFromDriver: m.updater,
+		})
+
+		ctx, cancel := context.WithTimeout(m.ctx, 1*time.Second)
+		instance.WaitForFirstFingerprint(ctx)
+		if instance.lastHealthState != drivers.HealthStateUndetected {
+			availDrivers = append(availDrivers, id.Name)
+		}
+		m.instances[id.Name] = instance
+		cancel()
+	}
+
+	m.logger.Debug("detected drivers", "drivers", availDrivers)
+	if len(skippedDrivers) > 0 {
+		m.logger.Debug("drivers skipped due to allow/block list", "skipped_drivers", skippedDrivers)
+	}
+
+	select {
+	case <-m.ctx.Done():
+	}
+}
+
+// Shutdown cleans up all the plugins
+func (m *manager) Shutdown() {
+	// Cancel the context to stop any requests
+	m.cancel()
+
+	// Go through and shut everything down
+	for _, i := range m.instances {
```

Comment:
I think we'll need to lock around instances since it's accessed in the Run goroutine and Shutdown goroutine.